src/main/scala/nodes/learning/KernelGenerator.scala

```diff
@@ -0,0 +1,273 @@
+package nodes.learning
+
+import breeze.linalg._
+import breeze.math._
+import breeze.numerics._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.SparkConf
+import org.apache.spark.SparkContext
+import org.apache.spark.SparkContext._
+import utils._
+
+
+// Trait to be implemented by Kernel functions
+trait KernelGenerator[T] {
```

Comment:
we have two choices (a) create a data structure called `KernelBlock` that stores two things a RDD and a DenseMatrix (which is K_block_block) (b) we throw it away and re-create it inside the algorithm given the KernelBlock. It shouldn't be very expensive and this was a performance optimization that doesn't hold up very well to abstraction